tuning/autotune.py

```diff
@@ -50,12 +51,36 @@ class CandidateTracker:
     unet_vmfb_hash: Optional[str] = None
     unet_benchmark_time: Optional[float] = None
 
+def fetch_available_devices() -> list[any]:
+    drivers = ireert.HalDriver.query()
+    all_device_ids = []
 
-def parse_devices(devices_str: str) -> list[int]:
-    """Parse a comma-separated list of device IDs (e.g., "1,3,5" -> [1, 3, 5])."""
-    devices = []
+    for driver_name in drivers:
+        try:
+            driver = ireert.get_driver(driver_name)
+            devices = driver.query_available_devices()
+            all_device_ids.extend(f"{driver_name}://{device['path']}" for device in devices)
+        except RuntimeError as e:
+            print(f"Could not initialize driver {driver_name}: {e}")
+
+    return all_device_ids
+
+def validate_devices(user_devices) -> None:
+    available_devices = fetch_available_devices()
+
+    for device in user_devices:
+        if device not in available_devices:
+            raise ValueError(f"Invalid device specified: {device}")
```

Comment:
Let's not use exceptions for something that's not really recoverable. Asserting with an error is fine `assert False, f"Invalid device ...")`

Reply:
I will use `handle_error(condition: bool, msg: str, level: int = logging.ERROR, error_type: Type[BaseException] = Exception, exit_program: bool = False)`